sklearn/cluster/_spectral.py

```diff
@@ -589,11 +589,11 @@ class SpectralClustering(ClusterMixin, BaseEstimator):
            Stella X. Yu, Jianbo Shi
            <https://www1.icsi.berkeley.edu/~stellayu/publication/doc/2003kwayICCV.pdf>`_
 
-    .. [4] `Toward the Optimal Preconditioned Eigensolver:
-           Locally Optimal Block Preconditioned Conjugate Gradient Method, 2001.
+    .. [4] :doi:`Toward the Optimal Preconditioned Eigensolver:
+           Locally Optimal Block Preconditioned Conjugate Gradient Method, 2001
            A. V. Knyazev
            SIAM Journal on Scientific Computing 23, no. 2, pp. 517-541.
-           <https://epubs.siam.org/doi/pdf/10.1137/S1064827500366124>`_
+           <10.1137/S1064827500366124>`
```

Comment:
I used the DOI for consistency with the links to the same article in this file. Also IIRC `make linkcheck` flag the epubs.siam.org link as broken

Reply:
Actually this DOI link does not lead to a downloadable PDF for me.

Maybe we should use https://www.researchgate.net/profile/Andrew-Knyazev/publication/2434555_Toward_The_Optimal_Preconditioned_Eigensolver_Locally_Optimal_Block_Preconditioned_Conjugate_Gradient_Method/links/5f1e060f299bf1720d67f9a7/Toward-The-Optimal-Preconditioned-Eigensolver-Locally-Optimal-Block-Preconditioned-Conjugate-Gradient-Method.pdf instead.

Reply:
The PDF link above those not directly link to the PDF. It redirects to an HTML page from which it's possible to download the PDF by clicking in a browser (probably with javascript enabled).

Maybe we should link to the HTML page then: https://www.researchgate.net/publication/2434555_Toward_The_Optimal_Preconditioned_Eigensolver_Locally_Optimal_Block_Preconditioned_Conjugate_Gradient_Method

Reply:
Actually, the following link directly lead to the PDF:

https://www.researchgate.net/profile/Andrew-Knyazev/publication/2434555_Toward_The_Optimal_Preconditioned_Eigensolver_Locally_Optimal_Block_Preconditioned_Conjugate_Gradient_Method/links/5f1e060f299bf1720d67f9a7/Toward-The-Optimal-Preconditioned-Eigensolver-Locally-Optimal-Block-Preconditioned-Conjugate-Gradient-Method.pdf?origin=publication_detail

Reply:
I suggest we have the DOI vs publically accessible PDF separately. This PR's goal is to get `make linkcheck` to run without errors.

Also this particular link (through the DOI) in used in multiple places:
```
❯ git grep 10.1137/S1064827500366124
sklearn/cluster/_spectral.py:           <10.1137/S1064827500366124>`
sklearn/cluster/_spectral.py:           <10.1137/S1064827500366124>`
sklearn/externals/_lobpcg.py:       pp. 517-541. :doi:`10.1137/S1064827500366124`
sklearn/externals/_lobpcg.py:           pp. 517-541. :doi:`10.1137/S1064827500366124`
sklearn/manifold/_spectral_embedding.py:      <10.1137/S1064827500366124>`
```